app/services/feature_flag.rb

```diff
@@ -66,12 +69,23 @@ def self.deactivate(feature_name)
   def self.active?(feature_name)
     raise unless feature_name.in?(FEATURES)
 
-    FEATURES[feature_name].feature.active?
+    feature_flag = FEATURES[feature_name]
+
+    return feature_flag.active unless feature_flag.active.nil?
+
+    feature_active = Redis.current.hget(REDIS_CACHE_KEY, feature_name)
+    if feature_active.nil?
+      feature_active = feature_flag.feature.active?.to_s
+      Redis.current.hset(REDIS_CACHE_KEY, feature_name, feature_active)
+    end
+    feature_flag.active = JSON.parse(feature_active)
   end
 
   def self.sync_with_database(feature_name, active)
     feature = Feature.find_or_initialize_by(name: feature_name)
     feature.active = active
+    FEATURES[feature_name].active = active
```

Comment:
I've been having a long think about this, in the current implementation we are reaching into the instance in a constant and amending it which I would avoid. This got me thinking, instead of caching the status of a single feature, why don't we cache all of them.

I've played around with this:
Adding the constants for the expiry and key:
```
CACHE_EXPIRES_IN = 1.day
FEATURE_FLAG_CACHE_KEY = 'feature-flags-status'.freeze
```
Then adding a method to build the cache key:
```
def self.cache_key
    CacheKey.generate(FEATURE_FLAG_CACHE_KEY)
  end
```
This cache key is generated from the feature flag latest updated at, meaning if a feature flag is updated from the console we don't risk serving stale data, we also don't need to sync with the database.

After that we can build our own key, value hash with the name to the activity and cache that as so:
```
def self.features_status
    Rails.cache.fetch(cache_key, expires_in: CACHE_EXPIRES_IN) do
      # TODO: This doesn't account for non persisted features, we need to merge the two before building
      # the new hash
      Feature.where(name: FEATURES.keys).each_with_object({}) do |feature, hash|
        hash[feature.name] = feature.active?
      end
    end
  end
```
This means we build our own status for the features without amending the constant, and with the latest values.

Finally we can now lookup the values if `active?` is called by the following:
```
def self.active?(feature_name)
    raise unless feature_name.in?(FEATURES)

    features_status[feature_name]
end
```
This solution needs refinement and tidy up as it doesn't address new features not yet persisted, but this way we avoid caching single statuses.

That said I am wondering around adding caching to the feature flags, although we make constant calls to the feature table per page load, these are very cheap, efficient and small queries, which take less than a millisecond. Would substituting those with a caching layer help at all as we would be making those to Redis instead, also we can't get away from at least one call to the database for the timestamps? Let me know your thoughts

Reply:
talked in person but for posterity: this does affect performance when deployed. Going to refine the code further (will post shortly) and test that

Reply:
@vigneshmsft @duncanjbrown I've pushed the changes, let me know your thoughts